multicluster/apis/multicluster/v1alpha1/tunnelendpoint_types.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TunnelEndpointSpec defines the desired state of TunnelEndpoint
+type TunnelEndpointSpec struct {
+	// Active shows the Gateway Node is active or not.
+	Active bool `json:"active,omitempty"`
+	// ClusterID of the member cluster.
+	ClusterID string `json:"clusterID,omitempty"`
+	// Name of Gateway Node
+	Name string `json:"name,omitempty"`
+	// PodCIDRs are the Pod CIDRs used by the member cluster.
+	PodCIDRs []string `json:"podCIDRs,omitempty"`
```

Comment:
Probably you missed my earlier question:
> When later we support active/active or active/standby, will all gateway instances share the same Subnets? If so, we should consider a CR that includes all gateways?

Reply:
yes, they will have the same subnets, I will check how to handle this, do you mean a Tunnelendpoint  CR with all external/internal IPs of all Gateway Nodes?

Reply:
How about let us think about how the active/active design will be like and decide based on that?

Reply:
E.g. how to do load balancing, how to implement failure detection and failover.

Reply:
Hi @jianjuns , here is my raw design to support active/active mode, Could you let me your suggestion? thanks

1. GatewayNode will be created automatically when user annotate a Node as GatewayNode
```
type GatewayNodeSpec struct {
	// ClusterID of the member cluster.
	ClusterID string `json:"clusterID,omitempty"`
	// Name of Gateway Node
	Name string `json:"name,omitempty"`
	// PodCIDRs are the Pod CIDRs used by the Node.
	PodCIDRs []string `json:"podCIDRs,omitempty"`
	// ServiceCIDR are the IP ranges used by Service ClusterIP.
	ServiceCIDR string `json:"serviceCIDR,omitempty"`
	// InternalIP of Gateway Node
	InternalIP string `json:"internalIP,omitempty"`
	// ExternalIP of Gateway Node
	ExternalIP string `json:"externalIP,omitempty"`
}
```
2. GatewayNodeExport will be created and updated automatically when a new GatewayNode is created/update. the name of GatewayNodeExport will the same as member cluster ID, and will be only one GatewayNodeExport includes all GatewayNodes.
```
type GatewayNodeExportSpec struct {
	// ClusterID of the member cluster.
	ClusterID string `json:"clusterID,omitempty"`
	// PodCIDRs are all Pod CIDRs used by the member cluster.
	PodCIDRs []string `json:"podCIDRs,omitempty"`
	// ServiceCIDR are the IP ranges used by Service ClusterIP.
	ServiceCIDR string `json:"serviceCIDR,omitempty"`
	// InternalIP of all Gateway Nodes
	InternalIP []string `json:"internalIP,omitempty"`
	// ExternalIP of all Gateway Nodes
	ExternalIP []string `json:"externalIP,omitempty"`
}


```
3. GatewayNodeImport will have the same spec as GatewayNodeExport and contains a GatewayNodeStatus slice which can be updated by background go routine in the Gateway Node Agent.

4. For failure detection and failover, I think we can run a background process to detect the connection of remote Gateway Node, if it fails in a given interval, it will mark corresponding GatewayNodeStatus as false, then, each Antrea agent will watch the change of GatewayNodeImport to add or remove the ExternalIP/InternalIP of GatewayNode in corresponding openflow rules.

Reply:
How we do load balancing between gateways? It is decided by the client or by the gateway or by users selected the gateway Nodes?

Reply:
I think for cross-cluster traffic, it will decided by Antrea agent in a Gateway Node. but there are actually two cases, one is the cross-cluster traffic from general node to two or more Gateway nodes, another case is cross-cluster traffic from Gateway Node to one or more remote Gateway Node, I was thinking we may use the way similar like Service ClusterIP load balancer, but I need take a look about the details to see if it's workable.